catkv: replace StoredCatalog with cached CatalogReader

postamar opened the pull request:
catkv: replace StoredCatalog with cached CatalogReader

This commit leverages the previous one and uses nstree.Catalog as the
data structure for caching all catalog data read by catkv.CatalogReader.
The StoredCatalog object is removed in favor of a cached CatalogReader
implementation.

This results in more consistent behavior when reading descriptors,
comments and zone configs.

Informs #88571.

Release note: None

----

nstree: refactor Catalog entries

This commit changes no functionality. Comments and zone configs are
joined by descriptor ID in the Catalog's by-ID btree entries.

Release note: None

Commit:
nstree: refactor Catalog entries

This commit changes no functionality. Comments and zone configs are
joined by descriptor ID in the Catalog's by-ID btree entries.

Release note: None

Commit:
catkv: replace StoredCatalog with cached CatalogReader

This commit leverages the previous one and uses nstree.Catalog as the
data structure for caching all catalog data read by catkv.CatalogReader.
The StoredCatalog object is removed in favor of a cached CatalogReader
implementation.

This results in more consistent behavior when reading descriptors,
comments and zone configs.

Informs #88571.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93104)
<!-- Reviewable:end -->

ajwerner reviewed (changes requested):
Did you have any interest in adding testing? Perhaps extending the datadriven testing in some way? 

> Reviewed 8 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93104#-:-NIdQqc49yYHu__rcSTi:b-r81yc9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @postamar)

___
*[`pkg/sql/catalog/nstree/by_id_map.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93104#-NIdQE6N6PiiG3iYwqIB:-NIdQE6N6PiiG3iYwqIC:b-ixbrey) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb6197b86fda43cecc5e9fdad427cd9d56219a38/pkg/sql/catalog/nstree/by_id_map.go#L40)):*
> ```Go
> func (t byIDMap) clear() {
> 	clear(t.t)
> 	btreeSyncPool.Put(t.t)
> ```

don't you ned to set t.t to be nil after this?
___
*[`pkg/sql/catalog/nstree/by_name_map.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93104#-NIdQ9n_BdTQwqm1snV5:-NIdQ9n_BdTQwqm1snV6:b-ixbrey) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb6197b86fda43cecc5e9fdad427cd9d56219a38/pkg/sql/catalog/nstree/by_name_map.go#L44)):*
> ```Go
> func (t byNameMap) clear() {
> 	clear(t.t)
> 	btreeSyncPool.Put(t.t)
> ```

don't you ned to set t.t to be nil after this?


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -81,10 +81,23 @@ type Collection struct {
+	// of descriptors, especially when attempting to reuse an old name: without
+	// this set the check that verifies that the target name is not already in
+	// use would systematically fail.
+	shadowedNames map[descpb.NameInfo]bool
```
nit: I understand that we never set the value to `false` for any key, so we don't need to worry about whether a key does not exist or the value is `false`. But I think the convention is use `struct{}` as the value type.

postamar replied:
Yes, this rings a bell. I'll adjust.

chengxiong-ruan replied:
a friendly reminder that this is not fixed. But I don't hold a strong opinion blocking this. Feel free to ignore.

postamar replied:
Sorry! Thanks will do.

postamar replied:
Done.

chengxiong-ruan commented on pkg/sql/catalog/descs/collection.go:
```diff
@@ -513,75 +543,28 @@ func (tc *Collection) WriteDesc(
+}
+
+// lookupDescriptorID returns the descriptor ID assigned to an object name.
+func (tc *Collection) lookupDescriptorID(
```
This is nice cleanup!

chengxiong-ruan reviewed (commented):
just one nit. this is much cleaner now. thanks for doing this.

chengxiong-ruan commented:
*[`pkg/sql/catalog/nstree/by_id_map.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93104#-NIdQE6N6PiiG3iYwqIB:-NIdbr0A2nFpK8rA5ngs:bevnqmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb6197b86fda43cecc5e9fdad427cd9d56219a38/pkg/sql/catalog/nstree/by_id_map.go#L40)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

don't you ned to set t.t to be nil after this?
</blockquote></details>

I think this is done by
```
dt.clear()
*dt = IDMap{}
```


<!-- Sent from Reviewable.io -->

postamar commented:
> Did you have any interest in adding testing? Perhaps extending the datadriven testing in some way?

I'd dismissed it early on without much thought as being "too difficult" but thinking about it some more, we could add some data-driven tests to `catkv` at this point, which assert on the results of the queries as well as the kv trace (the latter to test the caching behaviour).

> don't you ned to set t.t to be nil after this?

The method call is by value, so 🤷‍♂️

Commit:
catkv: add data-driven tests

This commit adds test coverage to the CatalogReader implementations.

Release note: None

postamar commented:
I have added some tests which I'm happy to take suggestions for! Ready for another look.

ajwerner commented on pkg/sql/catalog/internal/catkv/catalog_reader.go:
```diff
@@ -133,21 +120,55 @@ func (cr catalogReader) Codec() keys.SQLCodec {
 		b.Header.MaxSpanRequestKeys = 0
 		descsPrefix := catalogkeys.MakeAllDescsMetadataKey(codec)
 		b.Scan(descsPrefix, descsPrefix.PrefixEnd())
+		if isEventLoggingEnabled(ctx) {
```
Where'd you get this pattern? If you're using `VEventf` I don't think you need to wrap these in the conditional.

postamar replied:
The problem is that it `VEventf`ing unconditionally pollutes the event log and fails a bunch of tests which don't care about namespace lookups and other catalog operations.

ajwerner replied:
alrighty, fair enough. Another thing you could do is lift the check up so we don't need all these conditionals by assigning a closure above this function?

Like:

```go
eventf := log.VEventf
if !isEventLoggingEnabled() {
    evenf = func...
}
```

ajwerner reviewed (commented):
I think your approach to event logging is a lot. Just VEventf it and be done with it?

I didn't give this the most in-depth review, but the change seems fine to me.

postamar commented on pkg/bench/rttanalysis/testdata/benchmark_expectations:
```diff
@@ -1,57 +1,57 @@
-14,AlterTableAddForeignKey/alter_table_add_foreign_key_with_3_columns
+11,AlterRole/alter_role_with_1_option
+14,AlterRole/alter_role_with_2_options
+19,AlterRole/alter_role_with_3_options
```
Agreed. I'll look into it.

postamar replied:
For some reason my branch looks up system.role_options in the namespace table, which it shouldn't do because this should be cached. Something's not quite right.

postamar replied:
Turns out, we were losing the system database cache after every `Reset`. This is even better. Thanks!

postamar replied:
Fixed, and added a regression test. This testing framework is paying off already.

postamar commented:
Thanks. I looked at the rttanalysis changes more closely and found another performance bug: we were needlessly round-tripping to storage during queries involving non-existent IDs or names after a ScanAll, instead of giving up early because we already know that the ID or name doesn't exist.

postamar commented:
Alright, no more regressions now. I'll bors when CI gives the green light. Thanks for the reviews.

postamar commented:
bors r+

postamar commented:
bors r+
